host.go

```diff
@@ -0,0 +1,200 @@
+package f3
+
+import (
+	"bytes"
+	"context"
+	"time"
+
+	"github.com/filecoin-project/go-f3/gpbft"
+	"github.com/filecoin-project/go-f3/sim"
+	"golang.org/x/xerrors"
+)
+
+type Client interface {
+	gpbft.Signer
+	gpbft.Verifier
+	gpbft.Tracer
+
+	BroadcastMessage(context.Context, []byte) error
+	IncommingMessages() <-chan *gpbft.GMessage
+	Logger() Logger
+}
+
+type gpbftRunner struct {
+	client      Client
+	participant *gpbft.Participant
+	manifest    Manifest
+
+	selfMessageQueue chan *gpbft.GMessage //for the future when self messages are async
+
+	alertTimer *time.Timer
+
+	runningCtx context.Context
+	log        Logger
+
+	host *gpbftHost
+}
+
+type gpbftHost struct {
+	Client
+	runner *gpbftRunner
+}
+
+func newRunner(id gpbft.ActorID, m Manifest, client Client) (*gpbftRunner, error) {
+	runner := &gpbftRunner{
+		client:           client,
+		manifest:         m,
+		selfMessageQueue: make(chan *gpbft.GMessage, 20),
+		log:              client.Logger(),
+	}
+	host := &gpbftHost{
+		Client: client,
+		runner: runner,
+	}
+	runner.host = host
+
+	// create a stopped timer to facilitate alerts requested from gpbft
+	runner.alertTimer = time.NewTimer(100 * time.Hour)
+	if !runner.alertTimer.Stop() {
+		<-runner.alertTimer.C
+	}
+
+	runner.log.Infof("starting host for P%d", id)
+	p, err := gpbft.NewParticipant(id, host, gpbft.WithTracer(client))
+	if err != nil {
+		return nil, xerrors.Errorf("creating participant: %w", err)
+	}
+	runner.participant = p
+	return runner, nil
+}
+
+func (h *gpbftRunner) Run(ctx context.Context) error {
+	var cancel func()
+	h.runningCtx, cancel = context.WithCancel(ctx)
+	defer cancel()
+
+	// TODO(Kubuxu): temporary hack until re-broadcast and/or booststrap synchronisation are implemented
+	time.Sleep(2 * time.Second)
+
+	err := h.participant.Start()
+	if err != nil {
+		return xerrors.Errorf("starting a participant: %w", err)
+	}
+
+	messageQueue := h.client.IncommingMessages()
+loop:
+	for {
+		select {
+		case msg := <-h.selfMessageQueue:
+			_, err = h.participant.ReceiveMessage(msg, true)
+		default:
+		}
+		if err != nil {
+			break loop
+		}
+
+		select {
+		case <-h.alertTimer.C:
+			h.log.Infof("alarm fired")
+			err = h.participant.ReceiveAlarm()
+		case msg := <-h.selfMessageQueue:
+			_, err = h.participant.ReceiveMessage(msg, true)
+		case msg, ok := <-messageQueue:
+			if !ok {
+				err = xerrors.Errorf("incoming messsage queue closed")
+				break loop
+			}
+			_, err = h.participant.ReceiveMessage(msg, false)
+		case <-ctx.Done():
+			return nil
+		}
+		if err != nil {
```

Comment:
You'll need much better error handling here. Right now this will just bail upon receiving an invalid message. Wait for #264 to land, but more work needed.